### ISSUE TITLE
Return AST from low-level programmatic API [$10]

### ISSUE THREAD

**Issue by RReverser** [eslint/eslint#948]
Passing AST retrieved from different tools/preprocessors/bundlers would be really great as it would allow not to generate code and reparse output, but reuse already parsed tree which is interoperable by nature.

I believe everything that needs to be changed is here: https://github.com/eslint/eslint/blob/master/lib/eslint.js#L473-L482 (check if `typeof text === 'string'` and do everything as before and assign it directly to `ast` if it's object).

I'm aware of rules that require source code to be available, but those check formatting and code style only and so obviously do not apply for preprocessed AST (even in case when we auto-generate code from it before passing to `eslint`).

However, I'm not sure about rules that call `getTokenBefore` / `getTokenAfter` / etc., as most tools work on AST only and do not care about tokens array, so there may be mismatch between modified tree and original tokens. Should we (and can we) regenerate tokens array in that case from tree?

**Comment by IanVS** [comment]
I've posted a $10 bounty here.  Implementing this *may* simplify automatic configuration creation.

**Comment by willklein** [comment]
I don't think it would simplify the auto-configuration, but it would significantly improve its performance. Without this, we're repeatedly reading and parsing every file in a project, and doing that many times.

**Comment by ilyavolodin** [comment]
I don't think it will help with that either. Because while you might get AST from CLIEngine, you will not be able to pass it back to CLIEngine.

**Comment by willklein** [comment]
@ilyavolodin, if this and #1025 were both done, would that give us the means to pass the returned AST (and source) back in for reuse?

**Comment by ilyavolodin** [comment]
Yes, if both would be done, it would probably help speed things up.

**Comment by tqc** [comment]
I'm looking at implementing this based on changes made in #3217 to fix #1025. I can see a couple of possibilities for the API:

Make parse a public method

    var code = new SourceCode(text, linter.parse(text, config));
    doSomething(code.parseResult);
    linter.verify(code, config);
    
Add a method similar to the existing getAllComments

    linter.verify(text, config)
    var code = linter.getSourceCode()
    doSomething(code.parseResult);

Thoughts?

**Comment by nzakas** [comment]
I'm not sure I understand what problem this is solving. Can you explain?

**Comment by tqc** [comment]
@nzakas It's pretty much exactly what the title says - If you are using a tool alongside ESLint that requires an AST as input, this gives you access to the AST produced by ESLint rather than requiring you to generate a new one. Earlier comments suggest automatic configuration as a possible use case, and I can see some possibilities around automatically fixing errors where it would be quite useful.

Both of those could be accomplished with #1025 alone, but this makes it much more convenient by using the standard ESLint parse rather than requiring the caller to reimplement it.

Impact on the codebase should be fairly minimal - the first option at least can be done with only a single line change on top of #1025.

**Comment by BYK** [comment]
@nzakas - It would allow ESLint to be part of projects like https://github.com/asterjs/aster which I think is the future.

**Comment by IanVS** [comment]
I wonder, couldn't the public `parse` method of Espree just be used to get AST? 

As the [README](https://github.com/eslint/espree#usage) says:

```javascript
var espree = require("espree");
var ast = espree.parse(code);
```

What is the value of returning the AST from ESLint rather than straight from whatever parser you have configured ESLint to use (Espree, babel-eslint, etc.)?

**Comment by tqc** [comment]
@IanVS True, #1025 allows that much. However, the ESLint parse isn't exactly the same as the espree parse method - It includes things like finding the right parser to use and making sure that the result will include all the components needed by verify. The call to espree will end up being a lot longer than just parse(code).

**Comment by RReverser** [comment]
@IanVS The point is exactly not to parse code twice if ESLint does that anyway under the hood and you can just get and use it for further operations.

@BYK Thank you for mentioning aster! Didn't know someone still cares about that kind of projects ;)

**Comment by nzakas** [comment]
@BYK I'm not debating the usefulness, the issue has already been accepted. :)

@tqc I'm just trying to understand what you're suggesting because the two suggestions are very different in my mind. I think the second one is closer to what we want, but need to look more closely.

**Comment by tqc** [comment]
@nzakas Fair enough. I saw them as quite similar - both are three lines that can replace the existing call to verify if you want more control over the AST.

The main difference would be when you get access to the AST.

Calling parse, you get the AST first and could potentially use it to update the rule settings before calling verify. It also fits in quite naturally with #1025 - I like the idea of providing a parse method as a reference - if your custom parse function doesn't return the same sort of thing, you are doing it wrong.

Calling getSourceCode will only work after calling verify. It looks a bit cleaner than the first option in some ways, but I'm not sure if that limitation is worth it.

**Comment by RReverser** [comment]
Awesome, thanks @nzakas!
